### ISSUE TITLE
[release/2.14][B60] Four Inductor E2E models regress to fail_accuracy while 2.13 passed

### ISSUE THREAD

**Issue by mengfei25** [intel/torch-xpu-ops#4905]
## Summary

On B60, PyTorch release/2.14 XPU nightly shows four new Inductor E2E `fail_accuracy` results. The same cases passed with PyTorch release/2.13.

## Environment

- Hardware: B60
- Container: `intelgpu/ubuntu-26.04-rolling:26.18`
- Target PyTorch: `torch==2.14.0.dev20260805+xpu`
- Target torchvision: `torchvision==0.29.0.dev20260806+xpu`
- Target torchaudio: `torchaudio==2.11.0.dev20260806+xpu`
- Target Triton XPU: `triton-xpu==3.8.0+git1e2d42a0`
- Baseline: PyTorch release/2.13 XPU wheel, same tests passed
- Device: XPU
- Backend: Inductor
- CUDAGraphs: disabled
- Models are not included in pytorch CI or HUD

## Affected Cases

| Suite | Dtype | Mode | Model | Batch size | Result | Key failure |
|---|---|---:|---|---:|---|---|
| `timm_models` | `amp_bf16` | training | `fbnetv3_b` | 8 | `fail_accuracy` | `blocks.0.0.bn1.bias.grad` RMSE exceeds tolerance |
| `torchbench` | `float32` | inference | `detectron2_maskrcnn_r_50_fpn` | 4 | `fail_accuracy` | `pred_masks` / `instances` uint8 mismatch |
| `torchbench` | `float32` | inference | `detectron2_maskrcnn` | 4 | `fail_accuracy` | `pred_masks` / `instances` uint8 mismatch |
| `torchbench` | `bfloat16` | training | `timm_efficientnet` | 4 | `fail_accuracy` | `bn1.bias.grad` RMSE exceeds tolerance |

## Reproducer

```bash
docker run -it -e TZ=Asia/Shanghai \
  --device=/dev/mem --device=/dev/dri --group-add video --privileged \
  --shm-size=8g intelgpu/ubuntu-26.04-rolling:26.18 bash

curl -LsSf https://astral.sh/uv/install.sh | sh
source $HOME/.local/bin/env
uv venv myvenv --python 3.10 --clear
source myvenv/bin/activate
uv pip install pip wheel "setuptools<81"

uv pip install \
  torch==2.14.0.dev20260805+xpu \
  torchaudio==2.11.0.dev20260806+xpu \
  torchvision==0.29.0.dev20260806+xpu \
  triton-xpu==3.8.0+git1e2d42a0 \
  --index-url https://download.pytorch.org/whl/nightly/xpu --pre

pytorch_commit="$(python -c 'import torch; print(torch.version.git_version)')"
git clone https://github.com/pytorch/pytorch
cd pytorch
git checkout ${pytorch_commit}

uv pip install pip pandas psutil scipy requests
uv pip install -r .ci/docker/ci_commit_pins/huggingface-requirements.txt
uv pip install timm

git clone https://github.com/intel/torch-xpu-ops
rsync -avz ./torch-xpu-ops/.ci/benchmarks/ ./benchmarks/dynamo/

git clone https://github.com/pytorch/benchmark tmp-benchmark
cd tmp-benchmark
python install.py --continue_on_fail fbnetv3_b detectron2_maskrcnn_r_50_fpn detectron2_maskrcnn timm_efficientnet
export PYTHONPATH=${PWD}:${PYTHONPATH}
cd ..
```

Representative single-case commands:

```bash
python benchmarks/dynamo/timm_models.py -d xpu --accuracy --amp --amp-dtype bfloat16 \
  --training --backend=inductor --cold-start-latency -n10 \
  --disable-cudagraphs --only fbnetv3_b

python benchmarks/dynamo/torchbench.py -d xpu --accuracy --float32 \
  --inference --backend=inductor --cold-start-latency -n10 \
  --disable-cudagraphs --only detectron2_maskrcnn_r_50_fpn

python benchmarks/dynamo/torchbench.py -d xpu --accuracy --float32 \
  --inference --backend=inductor --cold-start-latency -n10 \
  --disable-cudagraphs --only detectron2_maskrcnn

python benchmarks/dynamo/torchbench.py -d xpu --accuracy --bfloat16 \
  --training --backend=inductor --cold-start-latency -n10 \
  --disable-cudagraphs --only timm_efficientnet
```

## Observed Results

```csv
dev,elapsed,suite,dtype,mode,name,scenario,batch_size,accuracy,calls_captured,unique_graphs,fallbacks_to_eager,graph_breaks,unique_graph_breaks,autograd_captures,autograd_compiles,cudagraph_skips,compilation_latency
xpu,187.811,timm_models,amp_bf16,training,fbnetv3_b,accuracy,8,fail_accuracy,459,2,0,5,4,0,0,0,68.486656
xpu,133.380,torchbench,float32,inference,detectron2_maskrcnn_r_50_fpn,accuracy,4,fail_accuracy,1246,73,4,51,5,0,0,0,69.747165
xpu,133.556,torchbench,float32,inference,detectron2_maskrcnn,accuracy,4,fail_accuracy,1246,73,4,51,5,0,0,0,68.819831
xpu,110.783,torchbench,bfloat16,training,timm_efficientnet,accuracy,4,fail_accuracy,239,1,0,5,4,0,0,0,36.638270
```

Representative failure messages:

```text
fbnetv3_b:
RMSE (res-fp64): 0.01330, (ref-fp64): 0.00232 and shape=torch.Size([16]). res.dtype: torch.float32, multiplier: 3.000000, tol: 0.010000
Accuracy failed for key name blocks.0.0.bn1.bias.grad

detectron2_maskrcnn_r_50_fpn / detectron2_maskrcnn:
Accuracy failed: uint8 tensor did not match
Accuracy failed for key name pred_masks
Accuracy failed for key name instances

timm_efficientnet:
RMSE (res-fp64): 0.00024, (ref-fp64): 0.00004 and shape=torch.Size([32]). res.dtype: torch.bfloat16, multiplier: 3.000000, tol: 0.001000
Accuracy failed for key name bn1.bias.grad
```

## Expected Behavior

These four cases should pass accuracy on B60 with PyTorch release/2.14, matching the PyTorch release/2.13 baseline result.

## Notes

- Two failures are training gradient accuracy mismatches in CNN models.
- Two failures are Detectron2 inference mask/instance mismatches in `uint8` outputs.
- `detectron2_maskrcnn_r_50_fpn` and `detectron2_maskrcnn` show the same captured graph counts and failure keys, so they may share the same underlying regression.

**Comment by bjarzemb** [comment]
I have this issue for `timm_efficientnet` during pre RC test with torch 2.14.0.dev20260805+xpu on BMG B580 Windows. Test passed in PT2.13

**Comment by weishi-deng** [comment]
fbnetv3_b: pass on B60 with 0805 nightly. Cannot reproduce.

**Comment by weishi-deng** [comment]
## Bisection results for regression (2.13→2.14)

**Root cause: [pytorch#190206](https://github.com/pytorch/pytorch/pull/190206)** — "Canonicalize reciprocal(sqrt(x)) to rsqrt(x), Update Batchnorm decomp"
- Commit: `d39a02cf8b`
- Introduced between `2026-07-18` and `2026-07-19` nightly wheels

### Affected models and verification

| Model | Original (0719) | After revert (0719) |
|-------|----------------|---------------------|
| `timm_efficientnet` (training, bf16) | fail | pass |
| `detectron2_maskrcnn_r_50_fpn` (inference, fp32) | fail | pass |
| `detectron2_maskrcnn` (inference, fp32) | fail | pass |

All 3 models share the same root cause.

### Fix: revert both changes together

#### 1. Post-grad peephole — disable
**File:** `torch/_inductor/fx_passes/post_grad.py`  
**Line:** `match.replace_by_example(repl, [x])` (inside `reciprocal_sqrt_to_rsqrt`)
```
    # DISABLED for XPU acc: match.replace_by_example(repl, [x])
```
This replaces `reciprocal(sqrt(x))` with `rsqrt(x)` globally, including training paths, where `rsqrt` has lower precision on XPU B60.

#### 2. BN decomp inference path — revert
**File:** `torch/_decomp/decompositions.py`, line ~2128
```diff
-        invstd = torch.rsqrt(running_var + eps)
+        invstd = 1 / torch.sqrt(running_var + eps)
```
This affects the eager reference, so inductor and eager need to match for accuracy checking.

**Both changes must be reverted together** — reverting either one alone still fails accuracy.

Reproduced on B60 with `torch==2.14.0.dev20260719+xpu` (confirmed pass on 0718, fail on 0719).

**Comment by weishi-deng** [comment]
fbnetv3_b: not reproduced on local PyTorch B60. Only reproduced in the container.  Needs more investigation.

**Comment by weishi-deng** [comment]
fbnetv3_b: caused by https://github.com/intel/torch-xpu-ops/commit/591d39fb1617367afc6f2b974f543b108e9e851d. With the cast flag, both failed on B60 with release 2.13 and 2.14 nightly. Without the cast flag, both pass. PR: https://github.com/intel/torch-xpu-ops/pull/5035

**Comment by daisyden** [comment]
<details>
<summary>label-issue: intel/torch-xpu-ops#4905</summary>

Trace mode: evidence-only from issue body (E2E accuracy report). Existing human labels preserved: `Accuracy`, `os: Windows`, `os: Linux`, `hw: BMG`.

| axis | value | reason |
|---|---|---|
| triage | `need_split` | 2 groups: BN-grad RMSE regression (training CNNs); detectron2 uint8 mask/instance mismatch (inference). |

## Group 1 — training BN-gradient accuracy regression (CNN models)

Root cause: BN gradient accuracy regression on B60 under Inductor training; `bn1.bias.grad` RMSE exceeds tolerance; passed on release/2.13, fails on release/2.14.

Analyzed case: timm_models fbnetv3_b amp_bf16 training (case 1 of 2).

| axis | value | reason |
|---|---|---|
| `type` | `Bug` | Reported accuracy regression; `issue_type` empty -> Bug. |
| `test` | `test: e2e` | `benchmarks/dynamo/timm_models.py --accuracy --backend=inductor`. |
| `module` | `module: inductor` | Inductor training accuracy regression (backend=inductor). |
| `priority` | `High` | Benchmark accuracy regression (fail_accuracy) that passed in 2.13, now fails without crash. |
| `os` | `os: Windows` | Preserved existing human label. |
| `os` | `os: Linux` | Preserved existing human label. |
| `hw` | `hw: BMG` | Preserved human label; B60 (Battlemage). |
| `dtype` | `dtype: amp_bf16` | Representative case is amp_bf16 training. |
| symptom | `Accuracy` | Human label + BN grad RMSE exceeds tolerance. |
| symptom | `regression` | Passed release/2.13, fails release/2.14. |
| symptom | `training` | Training mode gradient mismatch. |

Test cases (2):
- timm_models, amp_bf16, training, fbnetv3_b, bs=8 — `blocks.0.0.bn1.bias.grad` RMSE exceeds tolerance
- torchbench, bfloat16, training, timm_efficientnet, bs=4 — `bn1.bias.grad` RMSE exceeds tolerance

## Group 2 — detectron2 uint8 mask/instance mismatch (inference)

Root cause: Detectron2 mask/instance `uint8` output mismatch under Inductor inference on B60; passed on release/2.13, fails on release/2.14 (two detectron variants share graph counts/failure keys).

Analyzed case: torchbench detectron2_maskrcnn_r_50_fpn float32 inference (case 1 of 2).

| axis | value | reason |
|---|---|---|
| `type` | `Bug` | Reported accuracy regression; `issue_type` empty -> Bug. |
| `test` | `test: e2e` | `benchmarks/dynamo/torchbench.py --accuracy --backend=inductor`. |
| `module` | `module: inductor` | Inductor inference accuracy regression. |
| `priority` | `High` | Benchmark accuracy regression (fail_accuracy) that passed in 2.13, now fails without crash. |
| `os` | `os: Windows` | Preserved existing human label. |
| `os` | `os: Linux` | Preserved existing human label. |
| `hw` | `hw: BMG` | Preserved human label; B60 (Battlemage). |
| `dtype` | `dtype: float32` | Representative case is float32 inference. |
| symptom | `Accuracy` | uint8 tensor mismatch on pred_masks/instances. |
| symptom | `regression` | Passed release/2.13, fails release/2.14. |
| symptom | `inference` | Inference mode mask/instance mismatch. |

Test cases (2):
- torchbench, float32, inference, detectron2_maskrcnn_r_50_fpn, bs=4 — `pred_masks`/`instances` uint8 mismatch
- torchbench, float32, inference, detectron2_maskrcnn, bs=4 — `pred_masks`/`instances` uint8 mismatch

</details>

**Comment by daisyden** [comment]
<details>
<summary>label-issue: intel/torch-xpu-ops#4905</summary>

Trace mode: evidence-only from issue body (E2E accuracy report). Existing human labels preserved: `Accuracy`, `os: Windows`, `os: Linux`, `hw: BMG`.

| axis | value | reason |
|---|---|---|
| triage | `need_split` | 2 groups: BN-grad RMSE regression (training CNNs); detectron2 uint8 mask/instance mismatch (inference). |

## Group 1 — training BN-gradient accuracy regression (CNN models)

Root cause: BN gradient accuracy regression on B60 under Inductor training; `bn1.bias.grad` RMSE exceeds tolerance; passed on release/2.13, fails on release/2.14.

Analyzed case: timm_models fbnetv3_b amp_bf16 training (case 1 of 2).

| axis | value | reason |
|---|---|---|
| `type` | `Bug` | Reported accuracy regression; `issue_type` empty -> Bug. |
| `test` | `test: e2e` | `benchmarks/dynamo/timm_models.py --accuracy --backend=inductor`. |
| `module` | `module: inductor` | Inductor training accuracy regression (backend=inductor). |
| `priority` | `High` | Benchmark accuracy regression (fail_accuracy) that passed in 2.13, now fails without crash. |
| `os` | `os: Windows` | Preserved existing human label. |
| `os` | `os: Linux` | Preserved existing human label. |
| `hw` | `hw: BMG` | Preserved human label; B60 (Battlemage). |
| `dtype` | `dtype: amp_bf16` | Representative case is amp_bf16 training. |
| symptom | `Accuracy` | Human label + BN grad RMSE exceeds tolerance. |
| symptom | `regression` | Passed release/2.13, fails release/2.14. |
| symptom | `training` | Training mode gradient mismatch. |

Test cases (2):
- timm_models, amp_bf16, training, fbnetv3_b, bs=8 — `blocks.0.0.bn1.bias.grad` RMSE exceeds tolerance
- torchbench, bfloat16, training, timm_efficientnet, bs=4 — `bn1.bias.grad` RMSE exceeds tolerance

## Group 2 — detectron2 uint8 mask/instance mismatch (inference)

Root cause: Detectron2 mask/instance `uint8` output mismatch under Inductor inference on B60; passed on release/2.13, fails on release/2.14 (two detectron variants share graph counts/failure keys).

Analyzed case: torchbench detectron2_maskrcnn_r_50_fpn float32 inference (case 1 of 2).

| axis | value | reason |
|---|---|---|
| `type` | `Bug` | Reported accuracy regression; `issue_type` empty -> Bug. |
| `test` | `test: e2e` | `benchmarks/dynamo/torchbench.py --accuracy --backend=inductor`. |
| `module` | `module: inductor` | Inductor inference accuracy regression. |
| `priority` | `High` | Benchmark accuracy regression (fail_accuracy) that passed in 2.13, now fails without crash. |
| `os` | `os: Windows` | Preserved existing human label. |
| `os` | `os: Linux` | Preserved existing human label. |
| `hw` | `hw: BMG` | Preserved human label; B60 (Battlemage). |
| `dtype` | `dtype: float32` | Representative case is float32 inference. |
| symptom | `Accuracy` | uint8 tensor mismatch on pred_masks/instances. |
| symptom | `regression` | Passed release/2.13, fails release/2.14. |
| symptom | `inference` | Inference mode mask/instance mismatch. |

Test cases (2):
- torchbench, float32, inference, detectron2_maskrcnn_r_50_fpn, bs=4 — `pred_masks`/`instances` uint8 mismatch
- torchbench, float32, inference, detectron2_maskrcnn, bs=4 — `pred_masks`/`instances` uint8 mismatch

</details>
